src/Lykke.Common.Validation/ImageType/ImageTypeAttribute.cs

```diff
@@ -0,0 +1,113 @@
+    {
+        private readonly IDictionary<string, ImageType> _imageTypes = new Dictionary<string, ImageType>();
+
+        private static readonly IEnumerable<ImageType> DefaultImageTypesList = new List<ImageType>
+        {
+            ImageType.JpgImageType,
+            ImageType.PngImageType,
+            ImageType.GifImageType,
+            ImageType.BmpImageType
+        };
+
+        private static readonly IDictionary<string, ImageType> DefaultImageTypes = InitDefaultImageTypes();
+        
+        public ImageTypeAttribute(params string[] imageExtensions)
+        {
+            if (imageExtensions == null || imageExtensions.Length == 0)
+            {
+                _imageTypes = DefaultImageTypes;
+            }
+            else
+            {
+                foreach (var extension in imageExtensions)
+                {
+                    if (DefaultImageTypes.TryGetValue(extension, out var imageType))
+                    {
+                        _imageTypes.Add(extension, imageType);
+                    }
+                }
+            }
+        }
+
+        protected override ValidationResult IsValid(object value, ValidationContext validationContext)
+        {
+            var file = (IFormFile) value;
+
+            if (file == null)
+                return new ValidationResult("Value is null or empty.");
+
+
+            var extension = GetExtension(file.FileName);
+
+            if (!_imageTypes.TryGetValue(extension, out var imageType))
+                return new ValidationResult("File extension is invalid.");
+
+            using (var fileStream = file.OpenReadStream())
+            {
+                var longestSignature = imageType.HexSignatures.OrderByDescending(x => x.Length).FirstOrDefault();
+
+                if (longestSignature == null)
+                    return new ValidationResult("No signatures specified for type.");
+
+                var buffer = new byte [longestSignature.Length];
+                fileStream.Read(buffer, 0, buffer.Length);
+
+                if (!ValidateSignature(buffer, imageType.HexSignatures))
+                    return new ValidationResult("File is not an image, signature is invalid.");
+
+                return ValidationResult.Success;
+            }
+        }
+
+        private static bool ValidateSignature(byte[] file, IEnumerable<string> validHexSignatures)
+        {
+            if (file == null || validHexSignatures == null)
+                return false;
+
+            return validHexSignatures.Select(StringToByteArray)
+                .Any(byteSequence => file.Take(byteSequence.Length).SequenceEqual(byteSequence));
+        }
+
+        private static byte[] StringToByteArray(string hex)
+        {
+            var numberChars = hex.Length;
+            var bytes = new byte[numberChars / 2];
+            for (var i = 0; i < numberChars; i += 2)
+                bytes[i / 2] = Convert.ToByte(hex.Substring(i, 2), 16);
```

Comment:
Is it possible implement without Substring()? It produces too much temp objects

Reply:
Removed this method